### DISCUSSION TITLE
Monad implementations in V

### DISCUSSION THREAD

**Discussion by smartiniOnGitHub** [vlang/v#13031] [Ideas]
Hi all,
I'm doing some experiments writing some Monads in V, to check compiler stuff etc ...
Currently I have a very simple implementation of:
- Try (Failure/Success), for example see: https://www.scala-lang.org/files/archive/api/2.13.7/scala/util/Try.html
- Either , for example see: https://www.scala-lang.org/files/archive/api/2.13.7/scala/util/Either.html
- Optional (None/Some), see related in Scala ...
- others soon: Future, lazy values, etc ...

I tried to use even sum types but got compilation error (not sure at the moment they works with option types and generics), but it's not a problem; I could provide some failing example if useful.
I have some problems with interface/implementation casts that need to be further explored.
I'm having some trouble to handle generic block of code (in Scala I can pass to a function a generic block of code which is powerful but not sure it fits here), so for example at the moment I'm using a function for computing the value; I'm planning to do some checks even with closures and pointers to see if it becomes more generic.
I know that this is similar to what already V implements natively, but here I'm trying to check what is currently possible (a lot I have to say) and to expand features all in V code.

I've used this stuff years ago with Scala and it's powerful; but the best would be to integrate with some generic collections, like the ones just implemented under the 'datatypes' module; soon or later I'd like to add some immutable collections there but this is another story.

I'd be very happy to share with V guys; if interested I could draft a PR, or otherwise in a dedicated repo ...

What do you think ?
Regards

**Comment by igotfr** [comment]
@smartiniOnGitHub see the alternative Error Handling proposals to V:
https://github.com/vlang/v/discussions/12740
https://github.com/vlang/rfcs/issues/17

Type Checker:
https://github.com/vlang/rfcs/pull/7
https://github.com/vlang/rfcs/pull/7

**Comment by smartiniOnGitHub** [comment]
Hi @cindRoberta, thanks for the pointers; those discussions/proposals to me looks very interesting but probably too wide for my current analysis/sample implementation here; so probably it makes sense for my stuff to go in a dedicated repository (and check with V community for problems/improvements with my implementation) just to explore what is possible to achieve at the moment (and if possible integrate with a collections library or wrap V data structures, to see if there is some benefit); then maybe later see if it makes sense to move somewhere.

If useful, I'll post some update here as soon as I have some interesting result.
And at the same time, follow those proposals ;) .

Thanks for now. Regards

**Reply by igotfr** [reply]
you can too create your own modules in

https://vpm.vlang.io/
